server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -549,22 +549,17 @@ private ClusterState applyCreateIndexRequestWithExistingMetadata(final ClusterSt
                                                                                             throws Exception {
         logger.info("applying create index request using existing index [{}] metadata", sourceMetadata.getIndex().getName());
 
-        final Map<String, Map<String, Object>> mappings;
-        if (request.mappings().size() == 0) {
-            mappings = Collections.emptyMap();
-        } else {
-            assert request.mappings().size() == 1 : "expected source metadata mappings to have 1 type but it had: " + request.mappings();
-            String sourceMappings = request.mappings().values().iterator().next();
-            mappings = Collections.singletonMap(MapperService.SINGLE_MAPPING_NAME,
-                Collections.unmodifiableMap(MapperService.parseMapping(xContentRegistry, sourceMappings)));
+        if (request.mappings().size() > 0) {
```

Comment:
This new logic looks quite different. The existing logic didn't actually make sense to me:
* The assertion mentioned 'source metadata mappings', when what we're checking is the target mappings.
* Here we make sure to swap out the type name for the singleton type `_doc`. But then in `validateResize` we throw an error if we encounter any mapping besides `_default_`. So we necessarily throw an error if the mappings are not empty.